### PR TITLE
Feature: 🍰 add updated at to contributions

### LINKED ISSUE #2140: 🚀 [Feature] Updated At for Contributions
## 🚀 Feature
 - [x]  Add column `updated_at` to contributions
 - [x]  When a contribution (IN_PROGESS or PENDING) is updated, create a message with the previous memo text, previous contribution `month` and previous contribution amount for this contribution.
 - [x] Ensure that only contributions that are IN_PROGESS or PENDING can be edited
 - [x] `created_at` of message is:
   *  `created_at` of contribution if contribution has no `updates_at`
   *  old `updated_at` of contribution if `updated_at` is present 
 - [x] set new `updated_at` on contribution at the end of the db transaction
 
 Note: Do not put unchanged values into the message
 Note: Generate a text for a changed amount and append it to the message (or have it as only text if only the amount was changed)

### PR DESCRIPTION
**Pull request by Elweyn**
## 🍰 Pullrequest

As user I want to update a contribution that I send in, as admin I need to see the changes.
Add the old contribution memo to the messages with some more informations. 

### Issues

- fixes #2140 

### Todo

- [X] Add updated_at to the contributions table
- [x] When a contribution (IN_PROGESS or PENDING) is updated, create a message with the previous memo text, previous contribution month and previous contribution amount for this contribution.

### CODE REVIEW

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 1]
```diff
@@ -192,7 +194,14 @@ export class ContributionResolver {
       logger.error('user of the pending contribution and send user does not correspond')
       throw new Error('user of the pending contribution and send user does not correspond')
     }
-
+    if (
+      contributionToUpdate.contributionStatus !== ContributionStatus.IN_PROGRESS &&
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/ContributionResolver.ts:198]: use `contributionSattus.in(valueA, valueB)` ?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 2]
```diff
@@ -204,10 +213,38 @@ export class ContributionResolver {
 
     // all possible cases not to be true are thrown in this function
     validateContribution(creations, amount, creationDateObj)
+
+    const contributionMessage = ContributionMessage.create()
+    contributionMessage.contributionId = contributionId
+    contributionMessage.createdAt = contributionToUpdate.updatedAt
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/ContributionResolver.ts:219]: why is this needed? is updatedAt not set when the message is created?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 3]
```diff
@@ -204,10 +213,38 @@ export class ContributionResolver {
...
+      ? contributionToUpdate.updatedAt
+      : contributionToUpdate.createdAt
+    let changeMessage = ''
+    if (contributionToUpdate.memo !== memo) {
+      // TODO: Add message to changeMessage
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/ContributionResolver.ts:224]: what does this TODO mean?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 4]
```diff
@@ -204,10 +213,38 @@ export class ContributionResolver {
...
+      changeMessage += 'Alter Memo: ' + contributionToUpdate.memo + '\n'
+    }
+    if (contributionToUpdate.amount.toFixed(6) !== amount.toFixed(6)) {
+      // TODO: Add amount to changeMessage
+      changeMessage += 'Alter Betrag: ' + contributionToUpdate.amount + '\n'
+    }
+    const isDateDifferent =
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/ContributionResolver.ts:231]: why use a variable here? Do you reuse the value?

File: database/entity/0052-add_updated_at_to_contributions/Contribution.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+  @DeleteDateColumn({ unsigned: true, nullable: true, name: 'deleted_by' })
+  deletedBy: number
+
+  @OneToMany(() => ContributionMessage, (message) => message.contribution)
+  @JoinColumn({ name: 'contribution_id' })
+  messages?: ContributionMessage[]
+
+  @Column({ nullable: true, name: 'updated_at' })
+  updatedAt: Date
```
Comment (ulfgebhardt) [inline database/entity/0052-add_updated_at_to_contributions/Contribution.ts:94]: move this line above the messages relation, consider to put it before the deleted fields

File: database/migrations/0052-add_updated_at_to_contributions.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+export async function upgrade(queryFn: (query: string, values?: any[]) => Promise<Array<any>>) {
+  await queryFn(`ALTER TABLE \`contributions\` ADD COLUMN \`updated_at\` datetime DEFAULT NULL;`)
```
Comment (ulfgebhardt) [inline database/migrations/0052-add_updated_at_to_contributions.ts:5]: `ADD COLUMN ... AFTER ...`

**Review by ulfgebhardt — DISMISSED** [review]

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 2 above)
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:219]: See the issue. There is explained why, @ulfgebhardt

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 7]
```diff
@@ -204,10 +213,38 @@ export class ContributionResolver {
...
+
+    const contributionMessage = ContributionMessage.create()
+    contributionMessage.contributionId = contributionId
+    contributionMessage.createdAt = contributionToUpdate.updatedAt
+      ? contributionToUpdate.updatedAt
+      : contributionToUpdate.createdAt
+    let changeMessage = ''
+    if (contributionToUpdate.memo !== memo) {
+      // TODO: Add message to changeMessage
+      changeMessage += 'Alter Memo: ' + contributionToUpdate.memo + '\n'
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:225]: I suggest no text. Just display the whole contribution as it was before the update.

**Review by Mogge — CHANGES_REQUESTED** [review]
Please try to find a way, to display the whole content of the contribution as it was in the message, without any added text fragments.

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 4 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:231]: I did it cause I tested every expression in a console log.... Maybe we can use it for the logging later on

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 1 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:198]: This does not work, contributionStatus is a String and can't use the in method

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 3 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:224]: Was some remanent of the coding

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 2 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:219]: @ulfgebhardt & @Mogge
updatedAt is setten after the creation of the message, and the issue was saying createdAt of the message is updatedAt of the contribution if it has one else the createdAt.
After reflecting it I am not sure why and if it is not better to set the value of now on createdAt since the message is created now. If we let the logic so we could have this HISTORY message in between some old messages

File: database/migrations/0052-add_updated_at_to_contributions.ts (hunk 6 above)
Comment (ulfgebhardt) [reply database/migrations/0052-add_updated_at_to_contributions.ts:5]: `transactionId` like in the model? At least that seems the most intuitive to me.

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 7 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts]: Withdrew the textes

**Review by ulfgebhardt — APPROVED** [review]

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 8]
```diff
@@ -192,7 +194,14 @@ export class ContributionResolver {
       logger.error('user of the pending contribution and send user does not correspond')
       throw new Error('user of the pending contribution and send user does not correspond')
     }
-
+    if (
+      contributionToUpdate.contributionStatus !== ContributionStatus.IN_PROGRESS &&
+      contributionToUpdate.contributionStatus !== ContributionStatus.PENDING
+    ) {
+      throw new Error(
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:201]: Please log this error

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 9]
```diff
@@ -204,10 +213,33 @@ export class ContributionResolver {
...
+    let changeMessage = ''
+    if (contributionToUpdate.contributionDate.getTime() !== new Date(creationDate).getTime()) {
+      changeMessage += contributionToUpdate.contributionDate + '\n'
+    }
+    if (contributionToUpdate.memo !== memo) {
+      changeMessage += '---\n' + contributionToUpdate.memo + '\n'
+    }
+    if (contributionToUpdate.amount.toFixed(6) !== amount.toFixed(6)) {
+      changeMessage += '---\n' + contributionToUpdate.amount
+    }
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:231]:
```suggestion
    const changeMessage = `${creationDate}
---
${memo}
---
${amount.toFixed(6)}`
```
I suggest to always show the whole content, not only the changed. Why `toFixed(6)`.

**Review by Mogge — DISMISSED** [review]
